types/src/unit_tests/trusted_state_test.rs

```diff
@@ -0,0 +1,521 @@
+                assert_eq!(Some(latest_validator_set), expected_validator_set);
+            }
+            TrustedStateChange::Version {
+                new_state,
+                latest_li,
+            } => {
+                assert_eq!(new_state.latest_version(), expected_latest_version);
+                assert_eq!(latest_li, &expected_latest_li);
+            }
+        };
+    }
+
+    #[test]
+    fn test_ratchet_from_waypoint(
+        (_vsets, lis_with_sigs, latest_li) in arb_update_proof(
+            10,   /* start epoch */
+            123,  /* start version */
+            1..3, /* version delta */
+            1..3, /* epoch changes */
+            1..5, /* validators per epoch */
+        )
+    ) {
+        let first_epoch_change_li = lis_with_sigs.first().map(LedgerInfoWithSignatures::ledger_info).unwrap();
+        let waypoint = Waypoint::new(first_epoch_change_li)
+            .expect("Generating waypoint failed even though we passed an epoch change ledger info");
+        let trusted_state = TrustedState::from_waypoint(waypoint);
+
+        let expected_latest_version = latest_li.ledger_info().version();
+        let expected_latest_li = latest_li.clone();
+        let expected_latest_epoch_change_li = lis_with_sigs.last().cloned();
+        let expected_validator_set = expected_latest_epoch_change_li
+            .as_ref()
+            .and_then(|li_with_sigs| li_with_sigs.ledger_info().next_validator_set());
+
+        let change_proof = ValidatorChangeProof::new(lis_with_sigs, false /* more */);
+        let trusted_state_change = trusted_state
+            .verify_and_ratchet(&latest_li, &change_proof)
+            .expect("Should never error or be stale when ratcheting from waypoint with valid proofs");
+
+        match trusted_state_change {
+            TrustedStateChange::Epoch {
+                new_state,
+                latest_li,
+                latest_epoch_change_li,
+                latest_validator_set,
+            } => {
+                assert_eq!(new_state.latest_version(), expected_latest_version);
+                assert_eq!(latest_li, &expected_latest_li);
+                assert_eq!(Some(latest_epoch_change_li), expected_latest_epoch_change_li.as_ref());
+                assert_eq!(Some(latest_validator_set), expected_validator_set);
+            }
+            TrustedStateChange::Version { .. } =>
+                panic!("Ratcheting from a waypoint should always provide the epoch for that waypoint"),
+        };
+    }
+
+    #[test]
+    fn test_ratchet_version_only(
+        (_vsets, mut lis_with_sigs, latest_li) in arb_update_proof(
+            1,    /* start epoch */
+            1,    /* start version */
+            1..3, /* version delta */
+            1,    /* epoch changes */
+            1..5, /* validators per epoch */
+        )
+    ) {
+        // Assume we have already ratcheted into this epoch
+        let epoch_change_li = lis_with_sigs.remove(0);
```

Comment:
shall we assert `lis_with_sigs.length() == 1`? (or did I understand it incorrectly?)

Reply:
ahh, I'm intentionally generating with epoch_changes = 1 in this case. it will panic if that ever changed to zero for some reason.

Reply:
1 makes total sense. It’s just remove(0) Makes one think what’s left is still useful.   Maybe just comment we expect to see only one.

Reply:
1 makes total sense. It’s just remove(0) Makes one think what’s left is still useful.   Maybe just comment we expect to see only one.